packages/global-styles/src/renderer.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect } from '@wordpress/element';
+
+export function useRenderedGlobalStyles( styles = {} ) {
+	const generatedStyles = compileStyles( styles );
+
+	useEffect( () => {
+		// styleNodeId should match server ID, see global-styles.php
+		const styleNodeId = 'global-styles-inline-css';
+		let styleNode = document.getElementById( styleNodeId );
+
+		if ( ! styleNode ) {
+			styleNode = document.createElement( 'style' );
+			styleNode.id = styleNodeId;
+			document
+				.getElementsByTagName( 'head' )[ 0 ]
+				.appendChild( styleNode );
+		}
+
+		styleNode.innerText = generatedStyles;
+	}, [ generatedStyles ] );
+}
+
+function flattenObject( ob, token = '' ) {
+	const toReturn = {};
+
+	for ( const i in ob ) {
+		if ( ! ob.hasOwnProperty( i ) ) continue;
+
+		if ( typeof ob[ i ] === 'object' ) {
+			const flatObject = flattenObject( ob[ i ], token );
+			for ( const x in flatObject ) {
+				if ( ! flatObject.hasOwnProperty( x ) ) continue;
+
+				toReturn[ i + token + x ] = flatObject[ x ];
+			}
+		} else {
+			toReturn[ i ] = ob[ i ];
+		}
+	}
+	return toReturn;
+}
+
+function compileStyles( styles = {} ) {
```

Comment:
For reference, there's another place where we do this kind of thing https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/hooks/style.js#L30 We should look into consolidating somehow.